vision/use-cases/hey_llm/src/main.ts

```diff
@@ -0,0 +1,506 @@
+  const cached = cache?.get(cacheKey);
+  if (cached) {
+    return cached;
+  }
+  const oauthService = getGoogleService_();
+  let parentFolderID = PropertiesService.getDocumentProperties().getProperty(
+    PROP_KEY.DRIVE_FOLDER_ID,
+  );
+  if (!parentFolderID) {
+    parentFolderID = createDriveFolder_(oauthService);
+    PropertiesService.getDocumentProperties().setProperty(
+      PROP_KEY.DRIVE_FOLDER_ID,
+      parentFolderID,
+    );
+  }
+  const filename = sanitizeFileName(`${cacheKey}:${prompt.slice(0, 64)}.png`);
+  const driveUrl = checkDriveImage_(oauthService, filename, parentFolderID);
+  if (driveUrl) {
+    return driveUrl;
+  }
+  const pred = requestImagen_(oauthService, prompt, seed);
+  const thumbnailLink = uploadImageToDrive_(
+    oauthService,
+    pred.bytesBase64Encoded,
+    filename,
+    pred.mimeType,
+    parentFolderID,
+  );
+
+  // Trim resize parameters that comes after the equal mark.
+  const url =
+    thumbnailLink.indexOf('=') > -1
+      ? thumbnailLink.split('=').slice(0, -1).join('')
+      : thumbnailLink;
+  cache?.put(cacheKey, url);
+  return url;
+}
+
+/**
+ * A utility function to generate OAuth2 service.
+ * Copied from https://github.com/googleworkspace/apps-script-oauth2#1-create-the-oauth2-service
+ * @private
+ * @returns OAuth2 service
+ */
+function getGoogleService_() {
+  const clientID = PropertiesService.getDocumentProperties().getProperty(
+    PROP_KEY.OAUTH_CLIENT_ID,
+  );
+  const clientSecret = PropertiesService.getDocumentProperties().getProperty(
+    PROP_KEY.OAUTH_CLIENT_SECRET,
+  );
+  if (!clientID || !clientSecret) {
+    throw new Error(
+      'OAuth info is missing. Complete authorization from the "Extensions" menu',
+    );
+  }
+
+  // Create a new service with the given name. The name will be used when
+  // persisting the authorized token, so ensure it is unique within the
+  // scope of the property store.
+  return (
+    OAuth2.createService('google')
+
+      // Set the endpoint URLs, which are the same for all Google services.
+      .setAuthorizationBaseUrl('https://accounts.google.com/o/oauth2/auth')
+      .setTokenUrl('https://accounts.google.com/o/oauth2/token')
+
+      // Set the client ID and secret, from the Google Developers Console.
+      .setClientId(clientID)
+      .setClientSecret(clientSecret)
+
+      // Set the name of the callback function in the script referenced
+      // above that should be invoked to complete the OAuth flow.
+      .setCallbackFunction('authCallback')
+
+      // Set the property store where authorized tokens should be persisted.
+      .setPropertyStore(PropertiesService.getDocumentProperties())
+      .setCache(CacheService.getDocumentCache()!)
+      .setLock(LockService.getDocumentLock())
+
+      // Set the scopes to request (space-separated for Google services).
+      .setScope([
+        'https://www.googleapis.com/auth/drive',
+        'https://www.googleapis.com/auth/cloud-platform',
```

Comment:
Are there libraries for this Authentication?

Reply:
Yes, this part is mostly copied from the usage section in
https://github.com/googleworkspace/apps-script-oauth2